agent/cache/type.go

```diff
@@ -41,6 +46,14 @@ type FetchOptions struct {
 	// Timeout is the maximum time for the query. This must be implemented
 	// in the Fetch itself.
 	Timeout time.Duration
+
+	// LastResult is the result from the last successful Fetch and represents the
+	// value currently stored in the cache at the time Fetch is invoked. It will
+	// be nil on first call where there is no current cache value. There may have
+	// been other Fetch attempts that resulted in an error in the mean time. These
+	// are not explicitly represented currently. We could add that if needed this
+	// was just simpler for now.
+	LastResult *FetchResult
```

Comment:
I would make it clear in the doc comment here: this is a pointer... can I modify any fields? Should I be very careful to definitely NOT modify any fields? etc.

Reply:
Actually, this comment was important - it turns out that even though the FetchResult is a new struct each time, because it's using `interface{}` for `Value` and `State`, it's actually really hard to avoid modifying the state directly in the cache entry during a Fetch!

While that isn't thread-unsafe because we guarantee only one Fetch per entry concurrently, it is really error prone since you might update the state and then hit an error and/or not update the index that goes with it which is super unintuitive.

Even if you return a struct or other value type and not a pointer as the state, if it contains any pointers things go bad.

So I'll update this wording here, but also the CA implementation in this PR currently _does_ use a pointer which means we inadvertently update the cache entry directly.